### PR TITLE
Add gated channels_last memory format for CUDA train and predict

### PR DESCRIPTION
**Pull request by raimbekovm**
## summary

adds an opt-in `channels_last=True` arg that converts the model and input batches to the nhwc memory format on cuda, for both training (`_setup_train` + train loop, including validation batches) and prediction (`setup_model` + `preprocess`). the conversion is gated to cuda and defaults off: on mps the layout is not numerically reliable and on cpu it gives no gain, so both warn once and fall back to nchw. prediction is additionally restricted to native pytorch models (`format == "pt"`) — exported cuda runtimes (tensorrt/onnx) read `im.data_ptr()` against an nchw binding, so an nhwc-strided input would be misread as nchw and corrupt predictions.

measured on a tesla t4 (imgsz 640, batch 16, amp), coco8 mAP is identical (Δ 0.00000) while throughput improves ~1.2x for both inference and training:

| model | inference | training |
|---|---|---|
| yolo26 n/s/m | 1.20 / 1.20 / 1.19× | 1.23 / 1.22 / 1.21× |
| yolo11 n/s/m | 1.24 / 1.22 / 1.21× | 1.26 / 1.22 / 1.23× |

Deleted: nothing (new gated opt-in capability; no existing code to remove or relocate).

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

Adds optional `channels_last` memory formatting to accelerate CUDA Tensor Core training and inference for native PyTorch models. 🚀

### 📊 Key Changes

- Introduces the `channels_last` configuration option for training, prediction, and validation, disabled by default.
- Converts models, input batches, preprocessing outputs, and warmup tensors to channels-last format when enabled.
- Restricts the feature to CUDA devices and native PyTorch models, with warnings when used on CPU, MPS, SAM, or exported formats such as ONNX and TensorRT.
- Updates classification and YOLOE validation paths to preserve the selected memory format.
- Ensures saved checkpoints use the standard NCHW format for compatibility with existing model loading and fusion logic.
- Documents the new argument in the training and prediction reference tables. 📚
- Aligns warmup inputs with the selected layout to avoid unnecessary `torch.compile` recompilation.

### 🎯 Purpose & Impact

- May improve training and inference speed on CUDA GPUs with Tensor Cores, without changing model results. ⚡
- Provides a simple configuration switch: `channels_last: True`.
- Maintains compatibility across CPU, MPS, SAM, and exported runtimes by safely ignoring unsupported configurations.
- Reduces the risk of incorrect predictions from exported backends that require standard NCHW input layout.
- Checkpoint files remain broadly compatible because they are serialized in the conventional format.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @raimbekovm, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms that your CUDA `channels_last` feature is ready for review; an Ultralytics engineer will assist with feedback and integration. Please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for the new `channels_last` feature, including its CUDA-only behavior and expected usage.
- ✅ **Add Tests**: If applicable, include or update tests to cover the CUDA memory-format changes, device gating, warnings, and training or prediction behavior, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The native PyTorch train/predict integration is compact and correctly converts model weights and training inputs after preprocessing. However, the predictor gate currently applies the layout change to CUDA exported runtimes that consume raw NCHW buffers, which must be addressed before merge.

💬 Posted 1 inline comment

**Review by UltralyticsAssistant — DISMISSED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The implementation is clean: the new option is registered as a boolean config field, safely gated to CUDA (and native PyTorch for prediction), and both model and input memory formats are converted in the relevant execution paths. No verified code issues found.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The CUDA gating and native-PyTorch backend restriction are correctly placed. One performance issue remains: prediction warmup does not match the newly selected memory format, defeating warmup for compiled channels-last inference.

💬 Posted 1 inline comment

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The CUDA gating, model/input conversion, and compile warmup alignment are well placed for standard YOLO predictors and trainers. One supported native-PyTorch predictor family (SAM) bypasses the new base implementation, causing `channels_last=True` to be silently ignored.

💬 Posted 1 inline comment

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The CUDA gating and native-PyTorch predictor restriction are sensible. One performance gap remains: validation during training does not convert input batches to the selected memory format.

💬 Posted 1 inline comment

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The CUDA gating and train/validation conversion paths are coherent. One specialized prediction path, classification, bypasses the new base preprocessing conversion, so the opt-in does not provide its intended inference behavior there.

💬 Posted 1 inline comment

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The training-side conversion and validation propagation are wired consistently. One prediction-path issue prevents the native model itself from being converted because the call targets the non-recursive AutoBackend wrapper.

💬 Posted 1 inline comment

File: ultralytics/engine/predictor.py [hunk 1]
```diff
@@ -422,6 +426,17 @@ def setup_model(self, model, verbose: bool = True):
...
+        # on MPS, no CPU gain. Exported CUDA runtimes (TensorRT / ONNX io-binding) read im.data_ptr() against an NCHW
+        # binding, so feeding them an NHWC-strided input would be misread as NCHW and corrupt predictions.
+        self.channels_last = self.args.channels_last and self.device.type == "cuda" and self.model.format == "pt"
+        if self.args.channels_last and not self.channels_last:
+            LOGGER.warning(
+                f"'channels_last=True' applies only to native PyTorch models on CUDA, ignoring for "
+                f"format='{self.model.format}' on '{self.device.type}'."
+            )
+        if self.channels_last:
+            self.model.to(memory_format=torch.channels_last)
```
Comment (raimbekovm) [inline ultralytics/engine/predictor.py:439]: `AutoBackend` subclasses `nn.Module` and overrides `_apply` (autobackend.py:373-390) to forward tensor ops into `self.backend.model`. `nn.Module.to(memory_format=...)` dispatches through `self._apply(convert)`, so the call does recurse into the native model. Verified on `yolo26n.pt`:

- before `self.model.to(memory_format=torch.channels_last)`: first conv weight stride `(27, 9, 3, 1)`, `is_contiguous(channels_last)=False`
- after: stride `(27, 1, 9, 3)`, `is_contiguous(channels_last)=True`

The conversion also runs before `attempt_compile`, so the model weights are channels_last at compile time. No change needed here.

Comment (UltralyticsAssistant) [inline ultralytics/engine/predictor.py:439]: ⚠️ **HIGH**: `self.model` is an `AutoBackend`, but its `backend` is a `BaseBackend` rather than an `nn.Module`, so `AutoBackend.to(memory_format=...)` has no registered parameters or buffers to convert. For native PyTorch inference this leaves `backend.model` in contiguous NCHW while inputs are converted to channels-last, so the prediction path does not actually enable channels-last weights or achieve the intended optimization. Convert the underlying native model instead.

**Suggested change:**
```suggestion
            self.model.model.to(memory_format=torch.channels_last)
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The training path correctly converts the direct training model and its batches, including validation. Prediction has one blocking issue: the conversion is applied to the AutoBackend wrapper rather than its native PyTorch model, making the advertised inference optimization ineffective.

💬 Posted 1 inline comment

**Review by glenn-jocher — APPROVED** [review]
Reviewed in depth. The gate design is right: opt-in default-off, trainer gates on CUDA, predictor additionally on `format == 'pt'` with an accurate rationale (exported CUDA runtimes read `data_ptr()` against NCHW bindings), warn-once when ignored, and `compile` sets the in-repo precedent for a user-facing perf knob. Benchmarks (1.19–1.26x on T4 with AMP, ΔmAP 0.00000) are consistent with known NHWC tensor-core behavior. I verified the default CPU path is untouched and confirmed the one open bot thread is a false positive — `AutoBackend._apply` forwards `.to(memory_format=...)` into the wrapped `PyTorchBackend.model`, reproduced on yolo26n-seg.pt (first conv stride (27,9,3,1) → (27,1,9,3), `is_contiguous(channels_last)=True`) — so the author's rebuttal stands and the thread can be closed.

Pushed two fixes (a43f369d8):
1. **Serialize checkpoints NCHW** in `save_model`: `deepcopy(ema).half()` preserved NHWC strides, so a channels_last-trained `best.pt` would crash every released version's `.view()`-based fusion and hand third parties non-canonical artifacts. Normalizing at the serialization owner fixes all consumers at once (and keeps #25363, now merged, as pure hardening rather than a load-bearing dependency).
2. **Standalone `val channels_last=True`** was silently ignored (`self.channels_last = self.training and trainer.channels_last`); it now honors the same CUDA+pt gate as predict, or warns — verified the warn path on CPU.

Worth a follow-up experiment: since a channels_last first conv propagates NHWC anyway, model-only conversion might achieve the same speedup — if benchmarks confirm, ~5 of the 7 conversion sites plus the `warmup` kwarg could be deleted.

**Comment by UltralyticsAssistant** [comment]
🎉 Merged in #25362 — thank you @raimbekovm, with contributions from @glenn-jocher!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

This contribution brings optional `channels_last` support to native PyTorch training, prediction, and validation on CUDA, helping Tensor Core workloads run more efficiently while preserving compatibility across devices and exported formats. The careful handling of warmup inputs, validation paths, and standard NCHW checkpoint saving makes this a valuable performance improvement without changing model results. 🚀
